crates/sc-proof-of-time/src/pot_client.rs

```diff
@@ -0,0 +1,113 @@
+//! Consensus node interface to the clock master network.
+
+use crate::pot_state::PotProtocolState;
+use crate::utils::{get_consensus_tip_proofs, topic, PotGossip, LOG_TARGET};
+use crate::PotComponents;
+use futures::{FutureExt, StreamExt};
+use parity_scale_codec::Decode;
+use sc_network::PeerId;
+use sp_blockchain::{HeaderBackend, Info};
+use sp_consensus::SyncOracle;
+use sp_runtime::traits::Block as BlockT;
+use std::sync::Arc;
+use std::time::Instant;
+use subspace_core_primitives::PotProof;
+use tracing::{debug, error, info, warn};
+
+/// The PoT client implementation
+pub struct PotClient<Block: BlockT, Client, SO> {
+    gossip: PotGossip<Block>,
+    pot_state: Arc<dyn PotProtocolState>,
+    client: Arc<Client>,
+    sync_oracle: Arc<SO>,
+    chain_info_fn: Arc<dyn Fn() -> Info<Block> + Send + Sync>,
+}
+
+impl<Block, Client, SO> PotClient<Block, Client, SO>
+where
+    Block: BlockT,
+    Client: HeaderBackend<Block>,
+    SO: SyncOracle + Send + Sync + Clone + 'static,
+{
+    /// Creates the PoT client instance.
+    pub fn new(
+        components: PotComponents,
+        gossip: PotGossip<Block>,
+        client: Arc<Client>,
+        sync_oracle: Arc<SO>,
+        chain_info_fn: Arc<dyn Fn() -> Info<Block> + Send + Sync>,
+    ) -> Self {
+        let PotComponents {
+            protocol_state: pot_state,
+            ..
+        } = components;
+        Self {
+            gossip,
+            pot_state,
+            client,
+            sync_oracle,
+            chain_info_fn,
+        }
+    }
+
+    /// Starts the workers.
+    pub async fn run(self) {
```

Comment:
I'm not sure why clockmaker and non-clockmaker should have different logic in terms of being a PoT clients. Yes, one of them also happens to generate proofs, but the rest of the logic should be the same, right?

The thing I noticed here is that you call `get_consensus_tip_proofs` below and since I suspect this function will be called during startup before any proofs are available it'll just panic right away.

Reply:
That would be part of the consensus/bootstrap testing. Idea/expected sequence is: first bring up clock master bootstrap node(genesis farmer), let it build block 1 and some more blocks which have proofs. Then bring up other clock masters, node clients, which would sync from the tip. Then it should not panic. But yeah, as we test the integration, this might change more if we discover anything else.

Keeping them separate makes it easier to compare with spec.

Reply:
> Idea/expected sequence is: first bring up clock master bootstrap node(genesis farmer), let it build block 1 and some more blocks which have proofs.

Genesis farmer and clock master are completely independent entities the way I see it, why do you expect them to be the same?

> Then bring up other clock masters, node clients, which would sync from the tip.

I don't see why they can't start even before genesis farmer.

> Then it should not panic.

If code is panicking depending on order of starting of applications then the panic usage is incorrect. We need to statically guarantee that panic will not happen or prove that it is something so extraordinary and unrecoverable that panic is a reasonable option. Here I don't see either of those.

Reply:
I agree with Nazar: genesis farmer is independent of PoT. We need genesis farmer to produce block 1 and all clockmasters can start PoT independently knowing that block slot and the inputs. As described in the spec. IMO we don’t need PoT bootnode as a concept.

Reply:
We discussed this with Rahul earlier today and agreed, this will be taken care of in subsequent PRs.